### PR TITLE
KAFKA-15432:RLM Stop partitions should not be invoked for non-tiered storage topics

### PR DESCRIPTION
**Pull request by hudeqi**
As described from (jira)[https://issues.apache.org/jira/browse/KAFKA-15432]

### CODE REVIEW

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (divijvaidya) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: This could be moved after the if {} logic in next line and merged after `partitionsToStop.filterNot(sp => errorMap.contains(sp.topicPartition))`. This way you won't call a contains() operation per partition. 

Also:
1. we want to exclude reserved/internal topics as well.
2. we want to add back the warn that we talked about in https://github.com/apache/kafka/pull/14329#discussion_r1315675510 since we no longer expect stopPartitions to be called for non-tiered and reserved topic.

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (divijvaidya) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:641]: this comment should be updated now

**Review by divijvaidya — COMMENTED** [review]
Please add a test where you call ReplicaManager.stopPartition() with non-tiered topic and verify (using mockito) that rlm.stopPartition() is never called for those partitions.

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (hudeqi) [reply core/src/main/scala/kafka/server/ReplicaManager.scala:641]: updated.

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (hudeqi) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: The calculation of `tieredEnabledPartitions` may not be placed after if{}, because `logManager.asyncDelete` will remove the corresponding unifiedLog from `currentLogs` in `logManager`, causing the calculation of `tieredEnabledPartitions` to be incorrect. But I have merged the two filter conditions.
1. The calculation of `tieredEnabledPartitions` actually has excluded internal topics, because https://github.com/apache/kafka/blob/8f8ad6db384ce30e8a5d848d3ed826a3f7a54dfe/core/src/main/scala/kafka/log/UnifiedLog.scala#L1887
2. added.
@divijvaidya

**Review by divijvaidya — APPROVED** [review]
Looks good! Waiting for CI to succeed here.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (kamalcph) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java]: The `else` block should go into Line 384. Also, please update the log statement, the partition may not be internal:

```
if (topicIdByPartitionMap.containsKey(tp)) {
    ...
} else {
    LOGGER.warn("StopPartition call is not expected for partition: {}", tpId);
}  
```

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (kamalcph) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: For readability, can we split this line into two and rename the variable:

```
    val remotePartitionsToStop = partitionsToStop.filter(
      sp => logManager.getLog(sp.topicPartition).exists(unifiedLog => unifiedLog.remoteLogEnabled())
    )
```
Also, can we remove the comment? It's not clear. A cluster can have topics with and without tiered storage, we cannot assume them as internal.

or something in similar lines.

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (kamalcph) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: we can preserve the same check and use the `remotePartitionsToStop` instead of `partitionsToStop`:

```
// exclude the partitions with offline/error state
val partitions = remotePartitionsToStop.filterNot(sp => errorMap.contains(sp.topicPartition)).toSet.asJava
```

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (kamalcph) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Why `remoteStorageSystemEnable` is set to `true` by default. There is already one parameter `enableRemoteStorage` is available. Can we use it instead?

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (kamalcph) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:5692]: If remote storage is disabled, then we should not call `RLMOnLeadershipChange`.

**Review by kamalcph — CHANGES_REQUESTED** [review]
@hudeqi 

Left review comments to address. Thanks for working on this PR!

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (clolov) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java]: For my understanding why does deleteRemoteLog being false mean that the partition for which you are trying to stop the partition is a non-tiered or an internal topic? Can you not ask for a partition to be stopped if for example you are moving partitions around without asking for it to be deleted?

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (hudeqi) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java]: updated, thanks

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (hudeqi) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: updated.

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (hudeqi) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: updated.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (hudeqi) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Great! updated.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (hudeqi) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:5692]: The current logic is that regardless of whether the topic has tiered storage enabled, if `remoteLogManager` is not empty, `onLeadershipChange` will be called, but in this method, `filterPartitions` is used to filter out the topics that enable tiered storage. @kamalcph

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (hudeqi) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java]: Sorry, I misunderstood the meaning of adding warn log in the discussion, I have updated: https://github.com/apache/kafka/pull/14667#discussion_r1377397772

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (kamalcph) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java:380]: tp -> tpId

Also, remove the comments from L368-L370 since we are addressing it in this patch.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (kamalcph) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:3121]: Can we change the defaultTopicRemoteLogStorageEnable  value to `false` by default to preserve the same behaviour on other tests?

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (kamalcph) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: if (enableRemoteStorage && defaultTopicRemoteLogStorageEnable)

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (kamalcph) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]:
```suggestion
      val leaderAndIsr = LeaderAndIsr(0, 1, List(0, 1), LeaderRecoveryState.RECOVERED, LeaderAndIsr.InitialPartitionEpoch)
      val becomeLeaderRequest = makeLeaderAndIsrRequest(topicIds(tp0.topic), tp0, Seq(0, 1), leaderAndIsr)  
```

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (kamalcph) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:5692]: Got it. Thanks!

**Review by kamalcph — COMMENTED** [review]
Thanks for addressing the review comments! Left few more comments to address.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (hudeqi) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java:380]: `tpId` is `TopicIdPartition`, in the warn log, `topicIdByPartitionMap` not contains `tp`, how to get the `tpId`? @kamalcph

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (hudeqi) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:3121]: Default set true is to preserve the same behaviour on other tests. For example, for test `testStopReplicaWithDeletePartitionAndExistingPartitionAndEqualLeaderEpoch`, when `enableRemoteStorage` is set true, it will verify `verify(mockRemoteLogManager).stopPartitions(ArgumentMatchers.eq(Collections.singleton(stopPartition)), any())`, if `defaultTopicRemoteLogStorageEnable ` set false, in this pr change, the topic will be filtered out and test failed in the unit test, which is same as other unit tests. @kamalcph

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (hudeqi) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: updated

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (kamalcph) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java:380]: Then, we can keep it as `tp` itself.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (kamalcph) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:3121]: Let's update all the callers instead and supply the flag explicitly:

```
    val replicaManager = setupReplicaManagerWithMockedPurgatories(mockTimer, aliveBrokerIds = Seq(0, 1),
      enableRemoteStorage = enableRemoteStorage, defaultTopicRemoteLogStorageEnable = enableRemoteStorage)
```

The intention is that the default value should resemble the cluster default which is `false`.

**Review by kamalcph — APPROVED** [review]
LGTM, thanks for fixing this issue!

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (kamalcph) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:3121]: Went through the test cases, we can keep it as true.

**Comment by hudeqi** [comment]
Can this PR be merged? @divijvaidya

**Comment by divijvaidya** [comment]
I am restarting the CI. There are large number of test failures including some TS ones. Let's wait to have a saner CI build.

**Comment by hudeqi** [comment]
> I am restarting the CI. There are large number of test failures including some TS ones. Let's wait to have a saner CI build.

seems new CI failed tests are unrelated. @divijvaidya

**Comment by hudeqi** [comment]
> > I am restarting the CI. There are large number of test failures including some TS ones. Let's wait to have a saner CI build.
> 
> seems new CI failed tests are unrelated. @divijvaidya

The only failed test about TS: org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest is successfully in local env.

**Comment by divijvaidya** [comment]
All test failures are known to be flaky
```
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14667/9/testReport/junit/org.apache.kafka.trogdor.coordinator/CoordinatorTest/Build___JDK_17_and_Scala_2_13___testTaskRequestWithOldStartMsGetsUpdated__/)
[Build / JDK 8 and Scala 2.12 / kafka.server.DescribeClusterRequestTest.testDescribeClusterRequestExcludingClusterAuthorizedOperations(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14667/9/testReport/junit/kafka.server/DescribeClusterRequestTest/Build___JDK_8_and_Scala_2_12___testDescribeClusterRequestExcludingClusterAuthorizedOperations_String__quorum_kraft/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testFenceMultipleBrokers()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14667/9/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_8_and_Scala_2_12___testFenceMultipleBrokers__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14667/9/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_8_and_Scala_2_12___testBalancePartitionLeaders__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.tools.MetadataQuorumCommandTest.[5] Type=Raft-Combined, Name=testDescribeQuorumStatusSuccessful, MetadataVersion=3.7-IV1, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14667/9/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_8_and_Scala_2_12____5__Type_Raft_Combined__Name_testDescribeQuorumStatusSuccessful__MetadataVersion_3_7_IV1__Security_PLAINTEXT/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.tools.MetadataQuorumCommandTest.[1] Type=Raft-Combined, Name=testDescribeQuorumStatusSuccessful, MetadataVersion=3.7-IV1, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14667/9/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_8_and_Scala_2_12____1__Type_Raft_Combined__Name_testDescribeQuorumStatusSuccessful__MetadataVersion_3_7_IV1__Security_PLAINTEXT/)
[Build / JDK 21 and Scala 2.13 / integration.kafka.server.FetchFromFollowerIntegrationTest.testFetchFromFollowerWithRoll(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14667/9/testReport/junit/integration.kafka.server/FetchFromFollowerIntegrationTest/Build___JDK_21_and_Scala_2_13___testFetchFromFollowerWithRoll_String__quorum_kraft/)
[Build / JDK 21 and Scala 2.13 / org.apache.kafka.streams.integration.EosIntegrationTest.shouldWriteLatestOffsetsToCheckpointOnShutdown[at_least_once, processing threads = true]](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14667/9/testReport/junit/org.apache.kafka.streams.integration/EosIntegrationTest/Build___JDK_21_and_Scala_2_13___shouldWriteLatestOffsetsToCheckpointOnShutdown_at_least_once__processing_threads___true_/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationSSLTest.testReplicateSourceDefault()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14667/9/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationSSLTest/Build___JDK_11_and_Scala_2_13___testReplicateSourceDefault__/)
[Build / JDK 11 and Scala 2.13 / integration.kafka.server.FetchFromFollowerIntegrationTest.testRackAwareRangeAssignor(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14667/9/testReport/junit/integration.kafka.server/FetchFromFollowerIntegrationTest/Build___JDK_11_and_Scala_2_13___testRackAwareRangeAssignor_String__quorum_kraft/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest.testSendOffsetsToTransactionTimeout(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14667/9/testReport/junit/org.apache.kafka.tiered.storage.integration/TransactionsWithTieredStoreTest/Build___JDK_11_and_Scala_2_13___testSendOffsetsToTransactionTimeout_String__quorum_kraft/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.EosIntegrationTest.shouldWriteLatestOffsetsToCheckpointOnShutdown[exactly_once, processing threads = false]](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14667/9/testReport/junit/org.apache.kafka.streams.integration/EosIntegrationTest/Build___JDK_11_and_Scala_2_13___shouldWriteLatestOffsetsToCheckpointOnShutdown_exactly_once__processing_threads___false_/)
```

**Pull request merged**
